### PR TITLE
[Feat]FakeBaseModel for offline eagle; Kimi-K2.5 fixes;

### PR DESCRIPTION
**Pull request by h-guo18**
### What does this PR do?

Adds `FakeBaseModel` for offline EAGLE training and several Kimi-K2.5 compatibility fixes.

- **New**: `FakeBaseModel` — lightweight model that loads only `lm_head` and `embed_tokens` from a local checkpoint, avoiding full model weight loading during offline training. Configured via `FakeBaseArguments` and integrated into `load_vlm_or_llm`.
- **Fix**: `_find_base_model_parts` — support Kimi-K2.5 VLM layout (`language_model.model` path)
- **Fix**: offline mode lm_head access and CompressedTensors ignore path
- **Fix**: Kimi-K2.5 decoder `past_key_value`/`past_key_values` argument mismatch
- **Fix**: `rglob` for `.pt` discovery in nested offline data dirs; single-node GPU count respects `CUDA_VISIBLE_DEVICES`

Type of change: Bug fix, new feature

### Testing
Tested offline EAGLE training for Kimi-K2.5 end-to-end.

### Before your PR is "*Ready for review*"

Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/Model-Optimizer/blob/main/CONTRIBUTING.md) and your commits are signed (`git commit -s -S`).

Make sure you read and follow the [Security Best Practices](https://github.com/NVIDIA/Model-Optimizer/blob/main/SECURITY.md#security-coding-practices-for-contributors) (e.g. avoiding hardcoded `trust_remote_code=True`, `torch.load(..., weights_only=False)`, `pickle`, etc.).

- Is this change backward compatible?: ✅
- If you copied code from any other sources or added a new PIP dependency, did you follow guidance in `CONTRIBUTING.md`: N/A
- Did you write any new necessary tests?: ❌
- Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?: ❌

## Summary by CodeRabbit

* **New Features**
  * Lightweight fake-base model support for offline speculative-decoding training

* **Improvements**
  * Added CLI flags: --use_fake_base_for_offline, --trust_remote_code, and --fsdp
  * Expanded offline .pt discovery to include nested subdirectories
  * Better GPU detection with explicit single-node logging; FSDP enabled only when requested
  * Model loading and launch tooling now honor offline and trust-remote-code flags

* **Bug Fixes**
  * Improved compatibility with legacy transformer / Kimi-K2 call signatures

* **Tests**
  * Added tests covering fake-base loading and offline training workflows

### CODE REVIEW

File: modelopt/torch/speculative/utils.py [hunk 1]
```diff
@@ -443,6 +443,16 @@ def patched_fwd_with_lazy_rope_init(self, *args, **kwargs):
     kimi_k2_module.DeepseekV3Attention._init_rope = lambda self: None
     kimi_k2_module.DeepseekV3Attention.forward = patched_fwd_with_lazy_rope_init
 
+    # Kimi implementation is based on older transformers which use "past_key_value" argument
+    # We patch it to "past_key_values" for compatibility
+    original_decoder_layer_forward = kimi_k2_module.DeepseekV3DecoderLayer.forward
+
```
Comment (ChenhanYu) [inline modelopt/torch/speculative/utils.py:449]: `use_fake_base` defaults to `True` here, but `use_fake_base_for_offline` defaults to `False` in `ModelArguments`. Calling `load_vlm_or_llm(..., use_offline_training=True)` without explicitly passing `use_fake_base` would unexpectedly use `FakeBaseModel`. Default should be `False` to require explicit opt-in:
```python
def load_vlm_or_llm(
    model_name_or_path: str,
    use_fake_base: bool = False,  # not True
    ...
```

File: modelopt/torch/speculative/plugins/modeling_fakebase.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+        self.embed_tokens = nn.Embedding(config.vocab_size, config.hidden_size)
+        self.lm_head = nn.Linear(config.hidden_size, config.vocab_size, bias=False)
+
+        # Load lm_head and embed_tokens only from checkpoint
+        lm_head_w, embed_tokens_w = self._load_weights(source)
+        assert lm_head_w.shape == (config.vocab_size, config.hidden_size)
+        assert embed_tokens_w.shape == (config.vocab_size, config.hidden_size)
+        self.lm_head.weight.data.copy_(lm_head_w)
+        self.embed_tokens.weight.data.copy_(embed_tokens_w)
+
```
Comment (ChenhanYu) [inline modelopt/torch/speculative/plugins/modeling_fakebase.py:152]: The local vs remote loading paths duplicate ~20 lines of nearly identical code (index loading, key lookup, shard loading). Consider extracting a helper that resolves file paths first, then has a single loading path.

File: modelopt/torch/speculative/plugins/modeling_fakebase.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+    @staticmethod
+    def _find_weight_key(weight_map: dict, paths: list[str], label: str) -> str:
+        """Return the first ``path + '.weight'`` found in ``weight_map``."""
+        for path in paths:
+            key = path + ".weight"
+            if key in weight_map:
+                return key
+        tried = [p + ".weight" for p in paths]
+        raise RuntimeError(f"Cannot find {label} in checkpoint; tried: {tried}")
+
```
Comment (ChenhanYu) [inline modelopt/torch/speculative/plugins/modeling_fakebase.py:162]: Only safetensors checkpoints are supported (`model.safetensors.index.json`). Checkpoints using `pytorch_model.bin.index.json` or single-file formats will fail with a confusing `FileNotFoundError`. Either support them or raise a clear error stating safetensors is required.

File: modelopt/torch/speculative/utils.py [hunk 4]
```diff
@@ -474,21 +484,60 @@ def enable_cp_ttt_patch():
             modelopt.torch.speculative.plugins.transformers.ENABLE_CP_TTT_PATCH = False
```
Comment (ChenhanYu) [inline modelopt/torch/speculative/utils.py:484]: This copies `past_key_values` to `past_key_value` but doesn't remove `past_key_values` from kwargs. If the original forward has `**kwargs`, it could receive both. Use `pop` instead:
```python
def patched_decoder_layer_fwd(self, *args, **kwargs):
    kwargs["past_key_value"] = kwargs.pop("past_key_values", None)
    return original_decoder_layer_forward(self, *args, **kwargs)
```

File: modelopt/torch/speculative/plugins/transformers.py [hunk 5]
```diff
@@ -68,6 +67,7 @@
     get_ttt_msk_func,
     temporary_set_config_value,
```
Comment (ChenhanYu) [inline modelopt/torch/speculative/plugins/transformers.py:66]: The `CompressedTensorsConfig` ignore patch for Kimi-K2-Thinking was removed with no explanation. Was this addressed upstream? If not, removing it could re-introduce quantization of the drafter module. Please add a comment or commit message explaining why this is safe.

File: modelopt/torch/speculative/plugins/transformers.py [hunk 6]
```diff
@@ -464,19 +464,9 @@ def get_exporter(self) -> SpeculativeDecodingExporter:
     def _find_base_model_parts(self):
         """Find model parts from different models and set base_{part}_path attributes."""
         base_model_parts_mapping = {
-            "base_model_path": [
-                "model.language_model",
-                "model",
```
Comment (ChenhanYu) [inline modelopt/torch/speculative/plugins/transformers.py:480]: The attention mask `.repeat(batch_size, ...)` for kimi-k2 compatibility was removed. What changed to make this unnecessary? If this was a batch-size workaround, removing it could break multi-batch training.

File: tests/examples/speculative_decoding/test_eagle.py [hunk 7]
```diff
@@ -164,3 +203,68 @@ def test_convert_to_vllm_ckpt(tiny_llama_path, eagle_output_dir):
...
+@pytest.mark.parametrize(
+    ("model_source", "use_fake_base"),
+    [
+        (None, False),                       # tiny_llama (from fixture), no FakeBase
+        ("moonshotai/Kimi-K2.5", True),      # remote HF repo, FakeBaseModel
+        ("moonshotai/Kimi-K2-Thinking", True),     # remote HF repo, no FakeBaseModel
+        ("MiniMaxAI/MiniMax-M2.5", True),
+    ],
+    ids=["tinyllama", "kimi-k2.5","kimi-k2-thinking","minimax-m2.5"],
+)
```
Comment (ChenhanYu) [inline tests/examples/speculative_decoding/test_eagle.py:217]: `model_id` is hardcoded to `"kimi-k2.5"` for all non-tinyllama cases (kimi-k2-thinking, minimax-m2.5), so they all write to the same output directory and could interfere. Derive from `model_source`:
```python
model_id = model_source.split('/')[-1].lower() if model_source else "tinyllama"
```

File: tests/examples/speculative_decoding/test_eagle.py [hunk 8]
```diff
@@ -25,6 +26,44 @@
...
+    torch.manual_seed(42)
+    for i in range(num_files):
+        sample = {
+            "input_ids": torch.randint(0, vocab_size, (seq_len,)),
+            "hidden_states": torch.randn(seq_len, hidden_size),
+            "aux_hidden_states": torch.randn(seq_len, hidden_size * num_aux_layers),
+        }
+        torch.save(sample, output_dir / f"sample_{i:04d}.pt")
+    return output_dir
+
```
Comment (ChenhanYu) [inline tests/examples/speculative_decoding/test_eagle.py:65]: These 3 parametrized cases download from remote HF repos (potentially hundreds of GB). They should be marked `@pytest.mark.slow` or require a network marker so CI can skip them by default.

File: modelopt/torch/speculative/plugins/transformers.py (hunk 6 above)
Comment (h-guo18) [reply modelopt/torch/speculative/plugins/transformers.py:480]: We now use fake base model for kimi k2, so there is no need for this model-specific patch in main training code.

File: tests/examples/speculative_decoding/test_eagle.py (hunk 8 above)
Comment (h-guo18) [reply tests/examples/speculative_decoding/test_eagle.py:65]: we only download the shard that have lm head and embed_tokens, which is lower than 10 gb in total. This in  practice should take a few minutes only.

File: modelopt/torch/speculative/plugins/transformers.py (hunk 5 above)
Comment (h-guo18) [reply modelopt/torch/speculative/plugins/transformers.py:66]: we use fake base model now so no need for this patch.

**Comment by h-guo18** [comment]
> ## Review
> ### Issues
> **1. Security: `rglob("*.pt")` widens attack surface** (`eagle_utils.py:157`)
> 
> Switching from `glob` to `rglob` means recursively discovering `.pt` files from arbitrary subdirectories. Since `.pt` uses pickle, a malicious file in a nested dir could execute arbitrary code. Per SECURITY.md best practices, consider using `weights_only=True` in the corresponding `torch.load` call, or at least documenting the trust assumption.
> 
> **2. Shared path constants have new entries that change matching behavior** (`modeling_fakebase.py` → `transformers.py`)
> 
> The new shared lists add entries not present in `main`:
> 
> * `_BASE_MODEL_PATHS` adds `"language_model.model"` as the **first** entry
> * `_EMBED_TOKENS_PATHS` adds `"embed_tokens"` and `"language_model.model.embed_tokens"` at positions 0 and 1
> 
> Since `_find_base_model_parts` matches the **first** hit, the new ordering could cause incorrect matches for existing models. For example, a model that has both a top-level `embed_tokens` (unrelated) and the real one at `model.embed_tokens` would now match the wrong one. The original ordering in `main` was carefully chosen — verify the new order doesn't break existing models (Llama, Gemma, etc.).
> 
> **3. Missing `num_orig_hidden_layers` on FakeBaseModel path** (`utils.py:530-533`)
> 
> When `use_offline_training=True` and `use_fake_base=True`, the function returns `FakeBaseModel` early without setting `model.config.num_orig_hidden_layers`. The non-fake offline path (line 562) does set this. The EAGLE training loop in `main.py` previously relied on this attribute. `FakeBaseConfig` does set `num_hidden_layers` from the original config, but `num_orig_hidden_layers` is never set. If downstream code reads `num_orig_hidden_layers`, this will `AttributeError`.
> 
> **4. `load_vlm_or_llm_with_kwargs` removed without deprecation** (`utils.py`)
> 
> The old function is deleted outright. If there are any external consumers (scripts, notebooks, other repos), this is a breaking change. A one-line deprecation wrapper forwarding to `load_vlm_or_llm` would be low-cost and safer.
> 
> **5. Removed CompressedTensors quantization patch** (`transformers.py:577-580`)
> 
> The `CompressedTensorsConfig` ignore patch for the EAGLE module was removed without explanation. On `main`, this prevents the drafter from getting quantized when loading Kimi-K2-Thinking with compressed tensors. If that model is still supported with compressed tensors checkpoints, removing this could cause the EAGLE drafter to be quantized unexpectedly.
> 
> **6. Removed attention mask `.repeat()` for Kimi-K2** (`transformers.py:730-731`)
> 
> On `main`, `tensor_mask.repeat(batch_size, 1, 1, 1)` is present with the comment "repeat mask for kimi-k2 compatibility". The PR removes it. If Kimi-K2-Thinking still needs this, removing it will break that model's attention. The PR title says "Kimi-K2.5 fixes" — was this intentionally removed because K2.5 doesn't need it, or is K2-Thinking no longer supported?
> 
> **7. FSDP default behavior change is silent** (`launch_train.sh:206`)
> 
> On `main`, FSDP auto-enables for multi-GPU (`TOTAL_GPU > 1`). The PR changes this to require `--fsdp=True` explicitly. Existing users doing multi-GPU training without `--fsdp` will silently lose FSDP, potentially causing OOM or different training behavior. This should at minimum be documented in the PR description.
> 
> **8. `FakeBaseModel.__init__` signature is non-standard** (`modeling_fakebase.py:290`)
> 
> The constructor takes `source` instead of the standard `config` parameter expected by `PreTrainedModel`. This means `from_pretrained()`, `save_pretrained()`, and other HF utilities won't work. Fine if intentional, but worth a docstring note that this class is construction-only and not compatible with the standard HF save/load API.
> 
> **9. Tests download large remote models** (`test_eagle.py:238-241`)
> 
> The parametrized test downloads configs/indices from `moonshotai/Kimi-K2.5`, `moonshotai/Kimi-K2-Thinking`, and `MiniMaxAI/MiniMax-M2.5`. These will be slow and require network access. They should be gated (e.g., `@pytest.mark.slow` or an env var check) to avoid breaking fast CI runs.
> 
> ### Minor/Style
> * `modeling_fakebase.py`: License header says "Copyright (c) 2024" — should be 2025 for new files.
> * `test_eagle.py:641`: Missing spaces after commas in test IDs: `"kimi-k2.5","kimi-k2-thinking","minimax-m2.5"`.
> * `FakeBaseConfig.dtype` stores a `torch.dtype` object — this won't round-trip through HF's JSON config serialization cleanly.
> 
> ### What looks good
> * `FakeBaseModel` design is clean — loads only what's needed, auto-detects weight paths
> * `load_vlm_or_llm` with explicit parameters is much better API than `**kwargs`
> * The `self._base_model_lm_head` fix in offline mode is correct and important
> * Kimi decoder `past_key_value`/`past_key_values` patch is a clean compatibility fix
> * Single-node GPU detection via `torch.cuda.device_count()` respecting `CUDA_VISIBLE_DEVICES` is a good improvement
> * Removing hardcoded `trust_remote_code=True` in favor of a CLI flag is a security improvement
> * Unit tests are well-structured with proper monkeypatching

Comment 8 addressed in new commits. I think other comments does not make sense. Please let me know if you think they do. Thanks

File: modelopt/torch/speculative/utils.py [hunk 9]
```diff
@@ -474,21 +484,60 @@ def enable_cp_ttt_patch():
...
+        model_name_or_path,
+        trust_remote_code=trust_remote_code,
+        torch_dtype=torch_dtype,
+        device_map=device_map,
+        **extra,
+    )
+
+    if use_offline_training:
+        # Preserve the original layer count since we loaded with num_hidden_layers=0
+        model.config.num_orig_hidden_layers = model_config.num_hidden_layers
```
[the remark this answers is not in the document]
Comment (h-guo18) [reply modelopt/torch/speculative/utils.py:538]: k25 only supports fake base model workflow
